keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+This proposal adds an option to configure an endpoint IP family for a Kubernetes service:
+```
+    endpointFamily: <ipv4|ipv6|dual-stack>       [Default: dual-stack]
+```
+For example, the spec definition for an application that only binds to IPv4 might look like this:
+```
+spec:
+  selector:
+    app: MyApp
+  ports:
+  - protocol: TCP
+    port: 80
+    targetPort: 9376
+  endpointFamily: ipv4
+```
+This per-service endpoint IP family configuration is required for IPv4-only and IPv6-only Application Support. For example, if there is a legacy application that only binds to IPv4 addresses (IPv4-only application) that is running in an otherwise dual-stack cluster, then we don't want endpoints for the service to include IPv6 addresses that are allocated to backend pods for the service. Otherwise, any IPv6 addresses that are included in endpoints would be "dead ends" (no response will be received from the server) for ingress controller and external load balancer operation. The same reasoning applies to IPv6-only applications.
+
+If a service is configured for an endpointFamily of "ipv4" ("ipv6"), then endpoints for that port for the service will not include any IPv6 (IPv4) addresses that have been allocated to backend pods for the service.
+
+If a service is configured with an endpointFamily of "dual-stack", then both IPv4 and IPv6 endpoints will be created, but the service will only have a single service IP allocated (with a family that matches the cluster's configured service CIDR).
+
+If a service is exposed via nodePort, and the service is configured with an endpointFamily of "dual-stack", then iptables will be configured for both families, allowing both IPv4 and IPv6 forwarding.
+
+#### Modifying Consumers of EndpointAddress API to Use Dual-Stack Addresses
+Any clients that consume the [Kubernetes EndpointAddress API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpointaddress-v1-core) will need to be modified to read the new plural IPs field, if they are to take advantage of the dual-stack feature. This proposal will include dual-stack updates to the following consumers of Kubernetes endpoints:
+- kube-proxy, iptables mode
+- kube-proxy, IPVS mode
+- Nginx Ingress controller
+Other users/consumers of this API may be updated to dual-stack over time, but that work is considered outside of the scope of this proposal.
+
+#### 'kubectl get endpoints' Command Display for Dual-Stack Backend Pods
+The 'kubectl get endpoints ...' command display should be changed so that:
+- IPv4/IPv6 endpoint address pairs appear within curly braces.
+- If dual-stack endpoint address pairs are present, then each pair should be printed on its own line as show below.
+Example:
+```
+       kube-master# kubectl get endpoints
+       NAME            ENDPOINTS                             AGE
+       kubernetes      {[fd00::100]:6443,10.0.0.2:6643}      15m
+       nginx-service   {[fd00:db8:1::2]:80,192.168.1.3:80}   20m
```

Comment:
Do we have precedent for multi-line output like this?  @kubernetes/sig-cli-api-reviews